aspnetcore/mvc/controllers/routing.md

```diff
@@ -70,667 +65,788 @@ public class HomeController : Controller
+> In ASP.NET Core 3.0 and later, the routing system doesn't:
+> * Define a concept called a *route*. `UseRouting` adds route matching to the middleware pipeline. The `UseRouting` middleware looks at the set of endpoints defined in the app, and selects the best endpoint match based on the request.
+> * Provide guarantees about the execution order of extensibility like <xref:Microsoft.AspNetCore.Routing.IRouteConstraint> or <xref:Microsoft.AspNetCore.Mvc.ActionConstraints.IActionConstraint>.
+>
+>See [Routing](xref:fundamentals/routing) for reference material on routing.
 
-### Disambiguating actions
+<!-- review this section I added  -->
+<a name="cro"></a>
 
-When two endpoints match through routing, routing must disambiguate to choose the 'best' candidate or else throw an exception. For example:
+### Conventional routing order
 
-```csharp
-public class ProductsController : Controller
-{
-   public IActionResult Edit(int id) { ... }
+Conventional routing only matches a combination of action and controller that are defined by the app. This is intended to simplify cases where conventional routes overlap.
+Adding routes using <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapControllerRoute*>, <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapDefaultControllerRoute*>, and <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapAreaControllerRoute*> automatically assign an order value to their endpoints based on the order they are invoked. Matches from a route that appears earlier have a higher priority. Conventional routing is order-dependent. In general, routes with areas should be placed earlier as they're more specific than routes without an area. [Dedicated conventional routes](#dcr) with catch all route parameters like `{*article}` can make a route too [greedy](xref:fundamentals/routing#greedy), meaning that it matches URLs that you intended to be matched by other routes. Put the greedy routes later in the route table to prevent greedy matches.
 
-   [HttpPost]
-   public IActionResult Edit(int id, Product product) { ... }
-}
-```
+<a name="cctar"></a>
 
-This controller defines two actions that would match the URL path `/Products/Edit/17` and route data `{ controller = Products, action = Edit, id = 17 }`. This is a typical pattern for MVC controllers where `Edit(int)` shows a form to edit a product, and `Edit(int, Product)` processes  the posted form. To make this possible MVC would need to choose `Edit(int, Product)` when the request is an HTTP `POST` and `Edit(int)` when the HTTP verb is anything else.
+### ControllerContext.ToActionResult
+
+Many of the samples in this document use the following `ControllerContextExtensions`class to display <xref:Microsoft.AspNetCore.Mvc.Abstractions.ActionDescriptor> information:
```

Comment:
I think it's weird to show this code here because it deals with a bunch of APIs that are not described like `ControllerContext` and `ActionDescriptor`. Would it be appropriate to just link to the code?

Reply:
I would also *****REALLYY***** like to see this take a different shape than an extension method on a framework type called `ToActionResult`. This makes it really easy to forget that this isn't part of the product, I think it's artificially look built-in rather than make it clear that it's something we're using to explain the product.